fiftyone/utils/hf_hub.py

```diff
@@ -0,0 +1,527 @@
+            hf_split,
+            image_field,
+            download_dir,
+            split_name=split_name,
+        )
+        tmp_fo_dataset.compute_metadata()
+        for feature_name in hf_split.features:
+            if feature_name == image_field:
+                continue
+            _convert_hf_type_to_fiftyone(
+                tmp_fo_dataset, hf_split, feature_name
+            )
+        main_fo_dataset.merge_samples(tmp_fo_dataset)
+
+    if isinstance(hf_dataset, datasets.DatasetDict):
+        split_names = list(hf_dataset.keys())
+        for split_name in split_names:
+            print(f"Processing split '{split_name}'")
+            split = hf_dataset[split_name]
+            _add_split_to_fo_dataset(
+                fo_dataset,
+                split,
+                image_field,
+                download_dir,
+                split_name=split_name,
+            )
+    else:
+        _add_images(fo_dataset, hf_dataset, image_field, download_dir)
+        fo_dataset.compute_metadata()
+        _add_split_to_fo_dataset(
+            fo_dataset,
+            hf_dataset,
+            image_field,
+            download_dir,
+        )
+
+    return fo_dataset
+
+
+class BaseHuggingFaceLoader:
+    def __init__(self, repo_id: str, hf_dataset, **kwargs):
+        self.repo_id = repo_id
+        self.hf_dataset = hf_dataset
+        self.kwargs = kwargs
+
+    def load(self):
+        raise NotImplementedError
+
+
+class DefaultHuggingFaceLoader(BaseHuggingFaceLoader):
+    def __init__(self, repo_id: str, hf_dataset, **kwargs):
+        super().__init__(repo_id, hf_dataset, **kwargs)
+
+    def load(self):
+        return _convert_hf_dataset_to_fiftyone(
+            self.hf_dataset, self.repo_id, **self.kwargs
+        )
+
+
+def _is_convertable_to_fiftyone(repo_id):
+    api = hfh.HfApi()
+    return api.file_exists(repo_id, "fiftyone.py", repo_type="dataset")
+
+
+def _load_fiftyone_dataset_from_hub(repo_id, **kwargs):
+    download_dir = os.path.join(
+        fo.config.default_dataset_dir, "huggingface/hub", repo_id
+    )
+    hfh.snapshot_download(
+        repo_id=repo_id, repo_type="dataset", local_dir=download_dir
+    )
+
+    dataset = fod.Dataset.from_dir(
+        download_dir, dataset_type=fot.FiftyOneDataset, **kwargs
+    )
+    return dataset
+
+
+def _load_from_hub_with_script(repo_id, *args, **kwargs):
+    hf_dataset = datasets.load_dataset(repo_id, *args)
+    fo_module_path = hfh.hf_hub_download(
+        repo_id=repo_id, repo_type="dataset", filename="fiftyone.py"
+    )
+    conversion_module = _import_module_from_path(fo_module_path)
+    subclasses = _find_subclasses_in_module(
+        conversion_module, BaseHuggingFaceLoader
+    )
+    if len(subclasses) == 0:
+        loader = DefaultHuggingFaceLoader(repo_id, hf_dataset, **kwargs)
+    else:
+        loader = subclasses[0](repo_id, hf_dataset, **kwargs)
+    return loader.load()
+
+
+def load_from_hub(repo_id, *args, **kwargs):
```

Comment:
code readability suggestion: move `push_to_hub()` and `load_from_hub()` to the top of the module so anyone who wants to see what the module "does" will find the public API at the very top of the module.

Note that function declaration order does not matter in Python like it does in, say, C.